lib/puppet/property.rb

```diff
@@ -346,6 +362,21 @@ def insync?(is)
     end
   end
 
+  # This method tests if two values are insync? outside of the properties current
+  # should value. This works around the requirement for corrective_change analysis
+  # that requires two older values to be compared with the properties potentially
+  # custom insync? code.
+  #
+  # @param [Object] orig_should the original should value
+  # @param [Object] orig_is the original is value
+  # @return [Boolean] whether or not the values are in sync or not
+  # @api private
+  def insync_values?(orig_should, orig_is)
+    test_property = self.dup
```

Comment:
@adrienthebo @branan so here is the code in question. I've moved this into `property.rb` from `event.rb`, because its weird and weirdness like this probably belongs closer to the problem.

So ... following on from the previous PR comment ... this methodology provides results, but as you say you are concerned with the side-effects of a dup.

Let me begin by outlining our limitations. Today the insync? comparison is my best method for proving things are 'equal' or 'insync' after the fact. Without it, ensure properties won't match for example (say `present => absent` on a file fails because its really `file => absent`).

However the API of `#insync?` dictates you must set the `@should` instance variable (in this case via a `#should=` method) to do so. Unfortunately, `#insync?` is an abstract method often overwritten by custom types, and so an overridden method expects this as its API to work with. To point, we're limited to what we can change because it will break custom extensions.

As far as other ideas and implementations to achieve similar results ... I've incidentally tried another methodology, whereby I backed up the old value, do an insync? and restore the value, this at the time produced odd results that I didn't understand (as in, it was marking things out of sync when they were not). Unless you have a better idea, I can resurrect that methodology ^^ for investigation.